internal/controller/apiextensions/definition/composed.go

```diff
@@ -0,0 +1,256 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package definition
+
+import (
+	"context"
+	"strings"
+	"sync"
+
+	"github.com/google/uuid"
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	kunstructured "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kcache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	cache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	runtimeevent "sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/crossplane/crossplane-runtime/pkg/controller"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane/crossplane/internal/xcrd"
+)
+
+type composedResourceInformers struct {
+	log     logging.Logger
+	cluster cluster.Cluster
+
+	gvkRoutedCache *controller.GVKRoutedCache
+
+	lock      sync.RWMutex // everything below is protected by this lock
+	informers map[schema.GroupVersionKind]mrInformer
+	sinks     map[string]func(ev runtimeevent.UpdateEvent) // by some uid
+	xrs       map[string]xr                                // by controller name
+}
+
+type mrInformer struct {
+	cache    cache.Cache
+	cancelFn context.CancelFunc
+}
+
+type xr struct {
+	gvk   schema.GroupVersionKind
+	cache cache.Cache
+}
+
+var _ source.Source = &composedResourceInformers{}
+
+// Start implements source.Source, i.e. starting composedResourceInformers as
+// source with h as the sink of update events.
+func (i *composedResourceInformers) Start(ctx context.Context, h handler.EventHandler, q workqueue.RateLimitingInterface, ps ...predicate.Predicate) error {
+	id := uuid.New().String()
+
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.sinks[id] = func(ev runtimeevent.UpdateEvent) {
+		for _, p := range ps {
+			if !p.Update(ev) {
+				return
+			}
+		}
+		h.Update(ctx, ev, q)
+	}
+
+	go func() {
+		<-ctx.Done()
+
+		i.lock.Lock()
+		defer i.lock.Unlock()
+		delete(i.sinks, id)
+	}()
+
+	return nil
+}
+
+// RegisterXR maps a controller name to a GVK.
+func (i *composedResourceInformers) RegisterXR(name string, ca cache.Cache, gvk schema.GroupVersionKind) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	if i.xrs == nil {
+		i.xrs = make(map[string]xr)
+	}
+	i.xrs[name] = xr{
+		gvk:   gvk,
+		cache: ca,
+	}
+}
+
+// UnregisterXR removes a controller name to GVK mapping.
+func (i *composedResourceInformers) UnregisterXR(name string) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	delete(i.xrs, name)
+}
+
+func (i *composedResourceInformers) Watch(gvks ...schema.GroupVersionKind) {
```

Comment:
A comment explaining what this method does would help. To replay my own understanding, I think:

* We create a new cache for each GVK we're told we should watch (e.g. each kind of composed resource)
* In practice that cache will be backed by a single informer, for said GVK.
* We get the informer for said GVK.
* We add an event handler to the informer that sends events to all registered sinks (i.e. event handlers)
* In practice each event handler will trigger a reconcile of any XR controller that composes this GVK.
* We tell the routed cache to delegate operations on this GVK to this cache.

Reply:
done